kernel/aster-nix/src/fs/utils/dentrymnt.rs

```diff
@@ -0,0 +1,191 @@
+
+use super::{Dentry, FileSystem, InodeType, MountNode, NAME_MAX};
+use crate::prelude::*;
+
+/// The DentryMnt can represent a location in the mount tree.
+#[derive(Debug)]
+pub struct DentryMnt {
+    mount_node: Arc<MountNode>,
+    dentry: Arc<Dentry>,
+    this: Weak<DentryMnt>,
+}
+
+impl DentryMnt {
+    pub fn new(mount_node: Arc<MountNode>, dentry: Arc<Dentry>) -> Arc<Self> {
+        Arc::new_cyclic(|weak_self| Self {
+            mount_node,
+            dentry,
+            this: weak_self.clone(),
+        })
+    }
+
+    /// Lookup a dentrymnt.
+    pub fn lookup(&self, name: &str) -> Result<Arc<Self>> {
+        if self.dentry.inode().type_() != InodeType::Dir {
+            return_errno!(Errno::ENOTDIR);
+        }
+        if !self.dentry.inode().mode()?.is_executable() {
+            return_errno!(Errno::EACCES);
+        }
+        if name.len() > NAME_MAX {
+            return_errno!(Errno::ENAMETOOLONG);
+        }
+
+        let dentrymnt = match name {
+            "." => self.this(),
+            ".." => self.effective_parent().unwrap_or(self.this()),
+            name => {
+                let children_dentry = self.dentry.lookup_fast(name);
+                match children_dentry {
+                    Some(dentry) => Self::new(self.mount_node().clone(), dentry.clone()),
+                    None => {
+                        let slow_dentry = self.dentry.lookup_slow(name)?;
+                        Self::new(self.mount_node().clone(), slow_dentry.clone())
+                    }
+                }
+            }
+        };
+        let dentrymnt = dentrymnt.overlaid_dentrymnt();
+        Ok(dentrymnt)
+    }
+
+    // Get the absolute path.
+    //
+    // It will resolve the mountpoint automatically.
+    pub fn abs_path(&self) -> String {
+        let mut path = self.effective_name();
+        let mut dir_dentrymnt = self.this();
+
+        loop {
+            match dir_dentrymnt.effective_parent() {
+                None => break,
+                Some(parent_dir_dentrymnt) => {
+                    path = {
+                        let parent_name = parent_dir_dentrymnt.effective_name();
+                        if parent_name != "/" {
+                            parent_name + "/" + &path
+                        } else {
+                            parent_name + &path
+                        }
+                    };
+                    dir_dentrymnt = parent_dir_dentrymnt;
+                }
+            }
+        }
+        debug_assert!(path.starts_with('/'));
+        path
+    }
+
+    /// Get the effective name of dentrymnt.
+    ///
+    /// If it is the root of mount, it will go up to the mountpoint to get the name
+    /// of the mountpoint recursively.
+    fn effective_name(&self) -> String {
+        if !self.dentry.is_root_of_mount() {
+            return self.dentry.name();
+        }
+
+        if self.mount_node.parent().is_some() & self.mount_node.mountpoint_dentry().is_some() {
```

Comment:
Rewrite the big if-else block to the following form for improved efficiency and clarity. See my other comment about `effective_parent`.